### DISCUSSION TITLE
Guidance on preparing data for training from structures

### DISCUSSION THREAD

**Discussion by kirillshmilovich-insitro** [RMeli/gnina-torch#46] [Q&A] [answered]
Is there any info/tutorials/examples on how I can generate data files from my own structures/docked poses as input compatible for training gnina models. I know `libmolgrid` is used for handling the data to enable things random rotations for data augmentation, but I can't seem to find anything on how I can go about featurizing my own structures into the file scheme read by `libmolgrid`, (i.e., using the `ExampleProvider`). (edited)

**Comment by RMeli** [comment, marked as answer]
Unfortunately this is still very much WIP and there are no proper tutorials or instructions besides what you can find at [gnina-torch.rtfd.io](https://gnina-torch.rtfd.io/). 

However, you can find some examples on how to use `gninatorch.training` and `gninatorch.inference` under the `examples` folder.

As you say, `gninatorch` uses `libmolgrid` for grid-based featurisation of the molecular complexes, and you can find some  more information in the [libmolgrid documentation](https://gnina.github.io/libmolgrid/) and in the [libmolgrid tutorials](https://gnina.github.io/libmolgrid/tutorials.html). Essentially, what the `ExampleProvider.pupulate` expects is a file containing the labels, a path to the protein file, and a path to a ligand file:

```
0 protein1.pdb ligand1.sdf
1 protein1.pdb ligand2.sdf
0 protein1.pdb ligand3.sdf
...
```
This file is also the required argument of `gninatorch.training` and `gninatorch.inference`.

For molecular files, you can use whatever is supported by [Open Babel](https://openbabel.org/wiki/Main_Page), not just `.pdb` and `.sdf` files. `libmolgrid` takes care of I/O and you can then use `libmolgrid.gridmaker` to perform the featurisation. This is how `gninatorch` uses `libmolgrid` within `dataloaders.GriddedExamplesLoader`:

https://github.com/RMeli/gnina-torch/blob/34650b8b6d091631cfc50697cb72ab7371be8628/gninatorch/dataloaders.py#L112

https://github.com/RMeli/gnina-torch/blob/34650b8b6d091631cfc50697cb72ab7371be8628/gninatorch/dataloaders.py#L124-L140

If your data set is large, you might want to use [GNINA](https://github.com/gnina/gnina)'s `gninatyper` CLI utility to transform the  molecular files into `.types` files which are binary files containing only atom types and coordinates (i.e. the information `libmolgrid` needs for the featurisation). This will reduce I/O.

In `gninatorch.gnina` you can find a small example of how to set things up using `gninatorch`'s utilities, in case you want to use it within your own pipeline:
https://github.com/RMeli/gnina-torch/blob/34650b8b6d091631cfc50697cb72ab7371be8628/gninatorch/gnina.py#L321-L338

**Reply by kirillshmilovich-insitro** [reply]
Hey could you also clarify if `libmolgrid` generates voxel representations centered on the ligand? Or is there no centering performed?

**Reply by RMeli** [reply]
Yes, it does (but it depends on which `molgrid.GridMaker` method you use).

https://github.com/gnina/libmolgrid/issues/50

**Reply by kirillshmilovich-insitro** [reply]
Makes sense, but can you clarify the default behavior of the grid maker setup by `setup.setup_grid_maker` in gnina in this context? Do the voxelizations provided by `dataloaders.GriddedExamplesLoader` apply this centering by default?

**Reply by RMeli** [reply]
The voxelisations provided by `dataloaders.GriddedExamplesLoader` are indeed centered on the ligand (which should be the last coordinate set for each example in the `.types` file), before the optional random translations and rotations applied to the system.

`gninatorch.setup.setup_grid_maker()` is a very thin layer and it just returns a `molgrid.GridMaker`:
https://github.com/RMeli/gnina-torch/blob/34650b8b6d091631cfc50697cb72ab7371be8628/gninatorch/setup.py#L92-L94

In `dataloaders.GriddedExamplesLoader` the `GridMaker.forward()` function is called as follows:
https://github.com/RMeli/gnina-torch/blob/34650b8b6d091631cfc50697cb72ab7371be8628/gninatorch/dataloaders.py#L126-L133

Since `batch` is obtained with `molgrid.ExampleProvider.next_batch()`, this corresponds to the following signatures in [libmolgrid documentation](https://gnina.github.io/libmolgrid/python/index.html#molgrid.GridMaker.forward):

> forward((GridMaker)arg1, (ExampleVec)examplevec, (Grid5f)grid [, (float)random_translation=0.0 [, (bool)random_rotation=False]]) -> None :

> forward( (GridMaker)arg1, (ExampleVec)examples, (Grid5fCUDA)grid [, (float)random_translation=0.0 [, (bool)random_rotation=False]]) -> None :

**Reply by RMeli** [reply]
As you can see above, there was a small bug in the `random_translation` argument. This has been fixed in pre-release `0.0.2` (see #48).
